### PR TITLE
Add details for non-promotion causes/workarounds

### LINKED ISSUE #2940: Add promotion info/workaround to /null-safety
Details in https://github.com/dart-lang/sdk/issues/44900.

[Example from Lasse & mit](https://dartpad.dev/6e35cb85754ce3d44b1e94bf4d69caec?null_safety=true)

Note that `BadEvent` needs to use `year!` because `year` could have been changed by an overriding getter in a subclass. `GoodEvent` shadows `year` with a local variable (`var year = this.year`) so that the code can just use `year`.

### PR DESCRIPTION
**Pull request by kwalrath**
Fixes #2940

Staged: https://kw-staging-dartlang-2.firebaseapp.com/tools/non-promotion-reasons

### CODE REVIEW

**Comment by kwalrath** [comment]
@stereotype441 here's an early draft of how we might approach the no-promo page. The first case — only promoting local variables — seems like it might be the most common one by far, so I thought it might deserve its own section. Plus it seemed like a good example to feature (since we already did). But if people will always be dropping into individual sections, that doesn't make as much sense.

If people will mostly be dropping into individual sections, and if we don't have workarounds for each section, then we'll need a link back up to the common workaround section (https://kw-staging-dartlang-2.firebaseapp.com/tools/non-promotion-reasons#other-causes-and-workarounds).

Anyway, let me know what you think. I didn't want to get too far into this before talking with you.

/cc @bwilkerson

File: src/tools/non-promotion-reasons.md
Comment (stereotype441) [inline src/tools/non-promotion-reasons.md:32]: I'm guessing you got this example from me, because I write a lot test cases using `int` and `isEven` like this.  I wonder if it would be better to try to come up with less contrived examples since this is user-facing documentation.  I'm spitballing here, but maybe something like this?

```dart
class Order {
  DateTime? shipDate; // (1)
  void checkForShipmentDelay() {
    if (shipDate == null) return;
    if (DateTime.now().difference(shipDate) > Duration(days: 7)) { // (2) ERROR
      print('Shipment is late');
    }
  }
}
```

Unfortunately, I don't really have good ideas for making the other examples less contrived (they are less common than field promotion, so it's harder to invent concrete examples of where they might occur).  Maybe it's better to stick with `int` and `isEven` for now, and wait and see if anyone complains?  I don't know.

At the very least, consider changing this line to `print(i.isEven);` so that it doesn't just look like a statement with no effect 😃

File: src/tools/non-promotion-reasons.md
Comment (stereotype441) [inline src/tools/non-promotion-reasons.md:215]: Consider adding some placeholder text in the try block so that it looks more plausible that an exception might actually occur, e.g.:

```dart
try {
  i = j; // (1)
  ... other logic ...
  if (i == null) return; // (2)
  ... other logic ...
} catch (...) {
  print(i.isEven); // ERROR
}
```

File: src/tools/non-promotion-reasons.md
Comment (stereotype441) [inline src/tools/non-promotion-reasons.md:266]: Aha!  I just realized that the standard library already has a class that implements both `Comparable` and `Pattern`: the `String` class!  So if you want you could change this sentence to something like:

> Note that just because `Pattern` isn't a subtype of `Comparable` doesn't mean the code at (3) is dead; `o` could be a `String` (which implements both `Comparable` and `Pattern`).

**Review by stereotype441 — APPROVED** [review]
I like where you're going with this.  I added a few comments on details, but nothing really major.

File: src/tools/non-promotion-reasons.md
Comment (kwalrath) [reply src/tools/non-promotion-reasons.md:32]: Using print() would be nice. :) I do like keeping it simple, and contrived is fine as long as it helps readability/understandability.

**Comment by kwalrath** [comment]
What's your thinking on how people will reach this page? Do you expect it to mostly be through anchors (sections) or mostly to the top of the page? If we're unsure, I should probably link back up from each section to the common workaround section.

File: src/tools/non-promotion-reasons.md
Comment (kwalrath) [reply src/tools/non-promotion-reasons.md:32]: I just noticed a weird URL when I put the code in DartPad:

image

Comment (kwalrath) [reply src/tools/non-promotion-reasons.md:32]: (It should be http://dart.dev/go/non-promo-property.)

Comment (parlough) [reply src/tools/non-promotion-reasons.md:32]: This renders properly for me in WebStorm, so perhaps it's an issue with how DartPad sends or displays the diagnostic message.

Seems like it's received liked that from dart-services:
image

cc @domesticmouse

**Comment by kwalrath** [comment]
ptal @stereotype441. Can we give more guidance on how to fix each of these issues?

**Comment by kwalrath** [comment]
I've updated the staging server:

https://kw-staging-dartlang-2.firebaseapp.com/tools/non-promotion-reasons
https://kw-staging-dartlang-2.firebaseapp.com/go/non-promo-property
https://kw-staging-dartlang-2.firebaseapp.com/go/non-promo-this
https://kw-staging-dartlang-2.firebaseapp.com/go/non-promo-write

File: src/tools/non-promotion-reasons.md
Comment (parlough) [inline src/tools/non-promotion-reasons.md:77]: This tip fits in well here, I like it!

File: src/tools/non-promotion-reasons.md
Comment (stereotype441) [inline src/tools/non-promotion-reasons.md:366]: The best fix for this "particularly nasty" case is probably to make a fresh variable, e.g.:
```dart
void f(int? i) {
  var j = i ?? 0;
  var foo = () {
    print(j.isEven); // OK
  };
}
```
This works because `i ?? 0` has type non-nullable `int`; this will be used as the inferred type of `j`, so regardless of whether there's an assignment to `j` later, it can never take on a non-null value.

File: src/tools/non-promotion-reasons.md
Comment (stereotype441) [inline src/tools/non-promotion-reasons.md:398]: The best bet here is probably to do a separate variable, e.g.:
```dart
void f(int? i, int? j) {
  var foo = () {
    var i2 = i;
    if (i2 == null) return;
    print(i2.isEven); // OK; i2 can be promoted because it's local to this closure
  };
  var bar = () {
    i = j;
  };
}
```

**Review by stereotype441 — APPROVED** [review]
> ptal @stereotype441. Can we give more guidance on how to fix each of the problems (see my PENDING comments)?

Done!  I'm going to go ahead and mark this "approved" so that you don't have to wait for me if you want to land it as is, and iterate to make improvements later.  Or we can continue the conversation in this PR until it's fully polished - your call 😃

**Comment by kwalrath** [comment]
Great examples, @stereotype441. Thank you!

**Comment by kwalrath** [comment]
I've incorporated the examples but want to make another pass over it before asking for another review.

**Comment by kwalrath** [comment]
@stereotype441 I think we might be close to something I'd publish, but I'd like to figure out a few things:

* The right word(s) to use for "separate variable" / "fresh variable". I wonder if the word "copy" or "local" should be in there.
* Clarification on what makes `!` or `as` redundant.
* (could be later) The scope of this page, and when type promotion can and cannot occur. This is related to a discussion I had with @RedBrogdon today, regarding a question that @leafpetersen answered about type promotion via definite assignment.

In the meantime, I'll work on making the good examples be auto-included. We're close!

File: src/tools/non-promotion-reasons.md
Comment (bwilkerson) [inline src/tools/non-promotion-reasons.md:505]: I'll propose uniformly using a phase such as 'create a new variable', 'add a new variable', or even 'define a new variable' when a new variable is being introduced to solve the problem. For example: 'A solution is to create a new variable:'. Adding 'local' would also be reasonable: 'A solution is to create a new local variable:'.

**Review by bwilkerson — COMMENTED** [review]
> Clarification on what makes ! or as redundant.

It's redundant in the sense that the condition has already been checked but has to be checked again because inference wasn't able to use the result of the previous check.

File: src/tools/non-promotion-reasons.md
Comment (kwalrath) [reply src/tools/non-promotion-reasons.md:505]: I like the more specific one: "create a new local variable". Thanks, @bwilkerson!

Comment (kwalrath) [reply src/tools/non-promotion-reasons.md:505]: Upon further thought, the "new" seems redundant, so I'm thinking "create a local variable".

Comment (bwilkerson) [reply src/tools/non-promotion-reasons.md:505]: I agree. sgtm

**Comment by kwalrath** [comment]
I've created a .dart file to hold all the good examples, but I don't think the code's actually being analyzed. I'll look at that when I get back to this (Monday, I think).

The staging server is up-to-date:

https://kw-staging-dartlang-2.firebaseapp.com/tools/non-promotion-reasons

File: src/tools/non-promotion-reasons.md
Comment (parlough) [inline src/tools/non-promotion-reasons.md:259]: Perhaps another simplified anchor may be useful in this case as well.

**Review by parlough — APPROVED** [review]
Reads great  to me!

File: src/tools/non-promotion-reasons.md
Comment (kwalrath) [reply src/tools/non-promotion-reasons.md:259]: Good catch, I'll do that before merging.

**Pull request merged**
